chat-android/src/main/java/com/ably/chat/Typing.kt

```diff
@@ -119,130 +134,234 @@ internal class DefaultTyping(
+     * This does not prevent a client from sending an typing.stopped heartbeat—doing so resets the interval timer
+     * and allows another typing.started to be sent.
+     * Spec: CHA-T10
+     */
+    private val heartbeatThrottle: Duration = room.options.typing?.heartbeatThrottle ?: throw AblyException.fromErrorInfo(
+        ErrorInfo(
+            "Typing options hasn't been initialized",
+            ErrorCode.BadRequest.code,
+        ),
+    )
+
+    /**
+     * Defines how long to wait before marking a user as stopped typing.
+     * For example, if the last typing.started heartbeat was received 10s ago, the system will wait an additional 2s before declaring
+     * the client as no longer typing.
+     * @defaultValue 2 seconds
+     * Spec: CHA-T10a
+     */
+    private val timeout: Duration = 2.seconds
 
-    private var presenceSubscription: Subscription
+    private val eventBus = MutableSharedFlow<SelfTypingEvent>(extraBufferCapacity = kotlinx.coroutines.channels.Channel.UNLIMITED)
 
+    /**
+     * Spec: CHA-T13
+     */
     init {
+        // Send self typing events
         typingScope.launch {
-            eventBus.collect {
-                processEvent()
-            }
+            eventBus.collect { sendSelfTypingEvent(it) }
         }
 
-        val presenceListener = PresenceListener {
-            if (it.clientId == null) {
-                logger.error("unable to handle typing event; no clientId", context = mapOf("member" to it.toString()))
-            } else {
-                eventBus.tryEmit(Unit)
+        // Process received typing events
+        val typingListener = PubSubMessageListener { msg ->
+            val typingEventType = TypingEventType.entries.first { it.eventName == msg.name }
+            // CHA-T13a
+            if (msg.data == null || msg.data !is String) {
+                logger.error("unable to handle typing event; no clientId", context = mapOf("member" to msg.toString()))
+                return@PubSubMessageListener
+            }
+            typingScope.launch {
+                processReceivedTypingEvents(typingEventType, msg.data as String)
             }
         }
-
-        presenceSubscription = channelWrapper.presence.subscribe(presenceListener).asChatSubscription()
+        val typingEvents = listOf(TypingEventType.Started.eventName, TypingEventType.Stopped.eventName)
+        typingEventPubSubSubscription = channelWrapper.subscribe(typingEvents, typingListener).asChatSubscription()
     }
 
+    /**
+     * Spec: CHA-T6a
+     */
     override fun subscribe(listener: Typing.Listener): Subscription {
         logger.trace("DefaultTyping.subscribe()")
         listeners.add(listener)
+        // CHA-T6b
         return Subscription {
             logger.trace("DefaultTyping.unsubscribe()")
             listeners.remove(listener)
         }
     }
 
+    /**
+     * Spec: CHA-T9
+     */
     override suspend fun get(): Set<String> {
```

Comment:
In JS, we've made this function non-async, as we no longer have the problem of `presence.get()` being an async call - so we shouldn't need to do the `ensureAttached` call

Reply:
Thanks, will update 👍